### PR TITLE
Delve Debugger in Golang Lambda Container

### PR DESCRIPTION
**Pull request by austinlparker**
This change does the following:
- Adds `dlv` to the go lambda container image.
- Adds a 'debug' lambda mock that invokes the function via delve in headless mode

This change is in support of https://github.com/awslabs/aws-sam-cli/issues/281.

Please let me know if you'd prefer a different approach to this or have any other comments, thanks!

### CODE REVIEW

**Comment by mhart** [comment]
Hmmm, yeah the only thing I don't like is that there are extra binaries as `/var/lang/bin/dlv` and `/var/runtime/aws-lambda-debug-go`, which means this image deviates from production Lambda in terms of what's installed. (my main concern with that is ppl running code on docker-lambda that they think will run on a live Lambda)

That said, I'm not really sure what other options there are – when I was initially working on Go support, I asked @sanathkr, David Nasi and @bmoffatt about debugging approaches, including delve, but I don't think we got anywhere beyond "users will need to add it themselves".

Is there any way we can bundle delve together with the mock into a single `/var/runtime/aws-lambda-go` binary?

**Comment by austinlparker** [comment]
Hm, not sure. We could probably bring it in as an import to `aws-lambda-go` and try to launch it via the API. I'll hack on it a bit more and see what I can come up with.

**Comment by austinlparker** [comment]
It seems like it'd be pretty gnarly to bring in delve as a library and start the debugging service out of the `aws-lambda-go` binary. What I could do, though, is bring in just the `dlv` binary and make `aws-lambda-go` use a flag of some sort to debug the function that's passed in, rather than having a separate binary for it.

Another option would be to have a completely separate container tag - something like `go1.x-dbg` that included the additional binaries. That way there'd be a very clear line between the 'production' and 'non-production' version. Thoughts?

**Comment by austinlparker** [comment]
Alright, I spent a little more time poking at this problem and here's the three paths forward -

1. As-is; Have a single `go1.x` container image that includes delve and the debug wrapper.
2. Separate 'debug' container. This would involve creating a new `go1.xdebug` container image that would have `dlv` and the debug handler from this PR as part of the container image.
3. Mount delve into container from host. This is somewhat tricky because `dlv` needs to be recompiled for the container architecture but that's not the _worst_ thing in the world. In this scenario, the `dlv` binary would be removed and the `aws-lambda-go` binary could be modified to support launching the handler via the debugger.

@mhart thoughts on which option is preferable? I think the third is probably the most elegant and preserves the 'production-like' nature of the container.

**Comment by austinlparker** [comment]
I got into it and decided that the third one was the best and refactored things a bit.

Good:
- No new binaries added to the image.
- Doesn't break existing consumers; The flags to enable delve are implemented via `flags` and have sane defaults, and they don't break parsing of positional arguments.

Less Good:
- End-users will need to compile `dlv` themselves for `GOARCH=amd64 GOOS=linux` if they're running on MacOS or Windows.

I think the pros outweigh the cons.

**Comment by mhart** [comment]
Yeah, totally agree – that sounds like the best option. I'll look over it and see if any of the AWS ppl want to weigh in. 

@sanathkr ?

File: go1.x/run/aws-lambda-mock.go [hunk 1]
```diff
@@ -24,16 +25,21 @@ import (
 func main() {
 	rand.Seed(time.Now().UTC().UnixNano())
 
+	debugMode := flag.Bool("debug", false, "enables delve debugging")
+	delvePath := flag.String("delvePath", "/var/debug/dlv", "path to delve")
```
Comment (sanathkr) [inline go1.x/run/aws-lambda-mock.go]: `/var/debug` seems too generic. I don't want the container to break when official Lambda runtime suddenly makes use of this path in the future.

Instead, can you make it something explicit (ugly?) like `/tmp/lambci_debug_files`

File: go1.x/run/Dockerfile [hunk 2]
```diff
@@ -5,7 +5,6 @@ COPY aws-lambda-mock.go Gopkg.toml Gopkg.lock ./
 RUN dep ensure
 RUN GOARCH=amd64 GOOS=linux go build aws-lambda-mock.go
 
-
```
Comment (mhart) [inline go1.x/run/Dockerfile]: Can you revert this?

**Comment by austinlparker** [comment]
addressed all of the outstanding comments, squashed + pushed.

**Comment by mhart** [comment]
Awesome, merging :shipit:

**Comment by mhart** [comment]
Fixed in a4a89a532d4733fee2608b46959b7212bb8930b0 (I think)

**Comment by czubocha** [comment]
@austinlparker Delve flag `--api-version=1` was added for some particular reason? I suppose at the time of creating this PR VS Code did not support Delve API in version 2. But now VS Code supports both and default is 2 so i think it will be better to switch. I came here because running Delve with `--api-version=1` makes debugging Go Lambdas impossible at all in IntelliJ IDE (IDEA and Goland) as it uses version 2.

**Comment by austinlparker** [comment]
At the time VSC didn't support Delve API v2, so I forced it to 1, that's the only reason IIRC. I'm sure they'd accept a PR for changing it.

**Pull request merged**